third_party/proxy/healthcheck/proxier_health.go

```diff
@@ -0,0 +1,177 @@
+	healthTimeout time.Duration
+	recorder      events.EventRecorder
+	nodeRef       *v1.ObjectReference
+
+	lastUpdated atomic.Value
+	lastQueued  atomic.Value
+}
+
+// NewProxierHealthServer returns a proxier health http server.
+func NewProxierHealthServer(addr string, healthTimeout time.Duration, recorder events.EventRecorder, nodeRef *v1.ObjectReference) ProxierHealthUpdater {
+	return newProxierHealthServer(stdNetListener{}, stdHTTPServerFactory{}, clock.RealClock{}, addr, healthTimeout, recorder, nodeRef)
+}
+
+func newProxierHealthServer(listener listener, httpServerFactory httpServerFactory, c clock.Clock, addr string, healthTimeout time.Duration, recorder events.EventRecorder, nodeRef *v1.ObjectReference) *proxierHealthServer {
+	return &proxierHealthServer{
+		listener:      listener,
+		httpFactory:   httpServerFactory,
+		clock:         c,
+		addr:          addr,
+		healthTimeout: healthTimeout,
+		recorder:      recorder,
+		nodeRef:       nodeRef,
+	}
+}
+
+// Updated updates the lastUpdated timestamp.
+func (hs *proxierHealthServer) Updated() {
+	hs.lastUpdated.Store(hs.clock.Now())
+}
+
+// QueuedUpdate updates the lastQueued timestamp.
+func (hs *proxierHealthServer) QueuedUpdate() {
+	hs.lastQueued.Store(hs.clock.Now())
+}
+
+// Run starts the healthz HTTP server and blocks until it exits.
+func (hs *proxierHealthServer) Run() error {
+	serveMux := http.NewServeMux()
+	serveMux.Handle("/healthz", healthzHandler{hs: hs})
+	server := hs.httpFactory.New(hs.addr, serveMux)
+
+	listener, err := hs.listener.Listen(hs.addr)
+	if err != nil {
+		msg := fmt.Sprintf("failed to start proxier healthz on %s: %v", hs.addr, err)
+		// TODO(thockin): move eventing back to caller
+		if hs.recorder != nil {
+			hs.recorder.Eventf(hs.nodeRef, nil, "Warning", "FailedToStartProxierHealthcheck", "StartKubeProxy", msg)
+		}
+		return fmt.Errorf("%v", msg)
+	}
+
+	klog.V(3).InfoS("Starting healthz HTTP server", "address", hs.addr)
+
+	if err := server.Serve(listener); err != nil {
+		return fmt.Errorf("proxier healthz closed with error: %v", err)
+	}
+	return nil
+}
+
+type healthzHandler struct {
+	hs *proxierHealthServer
+}
+
+func (h healthzHandler) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
+	var lastQueued, lastUpdated time.Time
+	if val := h.hs.lastQueued.Load(); val != nil {
+		lastQueued = val.(time.Time)
+	}
+	if val := h.hs.lastUpdated.Load(); val != nil {
+		lastUpdated = val.(time.Time)
+	}
+	currentTime := h.hs.clock.Now()
+
+	healthy := false
+	switch {
+	case lastUpdated.IsZero():
+		// The proxy is healthy while it's starting up
+		// TODO: this makes it useless as a readinessProbe. Consider changing
+		// to only become healthy after the proxy is fully synced.
+		healthy = true
+	case lastUpdated.After(lastQueued):
+		// We've processed all updates
+		healthy = true
+	case currentTime.Sub(lastQueued) < h.hs.healthTimeout:
+		// There's an unprocessed update queued, but it's not late yet
+		healthy = true
+	}
```

Comment:
This seems incorrect. Even if `syncProxyRules` fails or becomes extremely slow, the proxy would be considered healthy as long as there are coming events if it compares with the last Queued timestamp. I think it should compare with the timestamp at which the first unprocessed event is received? 
I took a look at the latest version of kube-proxy, it has fixed it by https://github.com/kubernetes/kubernetes/commit/3fa4705aea4c713e15da7a6e2950db5e4f945e58